Add optional submitter parameter to FormData constructor

jenseng opened the pull request:
#### 136a23f12bae93a6dd2c9f93565aa6da6efdfef5
<pre>
Add optional submitter parameter to FormData constructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=251220">https://bugs.webkit.org/show_bug.cgi?id=251220</a>

Reviewed by Chris Dumez

Implement the new submitter parameter to the FormData constructor
Spec PR: <a href="https://github.com/whatwg/xhr/pull/366">https://github.com/whatwg/xhr/pull/366</a>
WPT PR: <a href="https://github.com/web-platform-tests/wpt/pull/37895">https://github.com/web-platform-tests/wpt/pull/37895</a>

Test: imported/w3c/web-platform-tests/xhr/formdata/constructor-submitter.html

* LayoutTests/imported/w3c/web-platform-tests/xhr/formdata/constructor-submitter-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/xhr/formdata/constructor-submitter.html: Added.
* Source/WebCore/html/DOMFormData.cpp:
(WebCore::DOMFormData::create): add submitter support
* Source/WebCore/html/DOMFormData.h: update signature
* Source/WebCore/html/DOMFormData.idl: update interface
* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::requestSubmit): improve error messages#

Canonical link: <a href="https://commits.webkit.org/259558@main">https://commits.webkit.org/259558@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/86174d63a2a68303d44e827354487f2fa10899ac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105032 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14111 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37922 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114295 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174478 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108935 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/15248 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5033 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97351 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/113308 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110788 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11783 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94786 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39297 "Found 1 new test failure: imported/w3c/web-platform-tests/fetch/stale-while-revalidate/stale-image.html (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93652 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26413 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80962 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7449 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27772 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7543 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4354 "Found 3 new test failures: fast/images/avif-as-image.html, fast/images/avif-heif-container-as-image.html, fast/images/background-image-relative-url-changes-document.html (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13598 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47324 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9332 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3520 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/xhr/formdata/constructor-submitter.html:
```diff
@@ -0,0 +1,92 @@
+<!DOCTYPE html>
```
In WebKit, you need to include the generated constructor-submitter-expected.txt as part of your patch (it gets generated when you run this test locally).

cdumez commented on Source/WebCore/html/DOMFormData.cpp:
```diff
@@ -32,6 +32,7 @@
 #include "DOMFormData.h"
 
 #include "Document.h"
+#include "HTMLElement.h"
```
Not needed.

cdumez commented on Source/WebCore/html/DOMFormData.cpp:
```diff
@@ -43,13 +44,24 @@ DOMFormData::DOMFormData(ScriptExecutionContext* context, const PAL::TextEncodin
+
+    RefPtr<HTMLFormControlElement> control;
+    if (submitter) {
+        if (!is<HTMLFormControlElement>(submitter))
```
Would look nicer like so:
```
if (submitter) {
    control = dynamicDowncast<HTMLFormControlElement>(*submitter);
    if (!control || !control->isSubmitButton())
        return Exception { TypeError, "The specified element is not a submit button."_s };
    if (control->form() != form)
        return Exception { NotFoundError, "The specified element is not owned by this form element."_s };
}
```

cdumez replied:
I fixed my code sample, it used to say `control->isSubmitButton()` instead of `!control->isSubmitButton()`.

jenseng replied:
Haha, good catch, I didn't even notice that, but my tests just caught it as well 😅

cdumez commented on Source/WebCore/html/HTMLFormElement.cpp:
```diff
@@ -355,12 +355,12 @@ ExceptionOr<void> HTMLFormElement::requestSubmit(HTMLElement* submitter)
     if (submitter) {
         // https://html.spec.whatwg.org/multipage/forms.html#dom-form-requestsubmit
         if (!is<HTMLFormControlElement>(submitter))
```
Can we refactor this code to look like what I suggested above too?

cdumez reviewed (approved):
LGTM

jenseng committed:
Add optional submitter parameter to FormData constructor
https://bugs.webkit.org/show_bug.cgi?id=251220

Reviewed by Chris Dumez

Implement the new submitter parameter to the FormData constructor
Spec PR: https://github.com/whatwg/xhr/pull/366
WPT PR: https://github.com/web-platform-tests/wpt/pull/37895

Test: imported/w3c/web-platform-tests/xhr/formdata/constructor-submitter.html

* LayoutTests/imported/w3c/web-platform-tests/xhr/formdata/constructor-submitter-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/xhr/formdata/constructor-submitter.html: Added.
* Source/WebCore/html/DOMFormData.cpp:
(WebCore::DOMFormData::create): add submitter support
* Source/WebCore/html/DOMFormData.h: update signature
* Source/WebCore/html/DOMFormData.idl: update interface
* Source/WebCore/html/HTMLFormElement.cpp:
(WebCore::HTMLFormElement::requestSubmit): improve error messages#

Canonical link: https://commits.webkit.org/259558@main
